File: includes/directory.php
<?php
/*
	FusionPBX
	Version: MPL 1.1

	The contents of this file are subject to the Mozilla Public License Version
	1.1 (the "License"); you may not use this file except in compliance with
	the License. You may obtain a copy of the License at
	http://www.mozilla.org/MPL/

	Software distributed under the License is distributed on an "AS IS" basis,
	WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
	for the specific language governing rights and limitations under the
	License.

	The Original Code is FusionPBX

	The Initial Developer of the Original Code is
	Mark J Crane <markjcrane@fusionpbx.com>
	Portions created by the Initial Developer are Copyright (C) 2008-2012
	the Initial Developer. All Rights Reserved.

	Contributor(s):
	Mark J Crane <markjcrane@fusionpbx.com>
*/
include "root.php";
require_once "includes/require.php";
require_once "includes/checkauth.php";

if (!function_exists('phone_letter_to_number')) {
	function phone_letter_to_number($tmp) {
		if ($tmp == "a" | $tmp == "b" | $tmp == "c") { return 2; }
		if ($tmp == "d" | $tmp == "e" | $tmp == "f") { return 3; }
		if ($tmp == "g" | $tmp == "h" | $tmp == "i") { return 4; }
		if ($tmp == "j" | $tmp == "k" | $tmp == "l") { return 5; }
		if ($tmp == "m" | $tmp == "n" | $tmp == "o") { return 6; }
		if ($tmp == "p" | $tmp == "q" | $tmp == "r" | $tmp == "s") { return 7; }
		if ($tmp == "t" | $tmp == "u" | $tmp == "v") { return 8; }
		if ($tmp == "w" | $tmp == "x" | $tmp == "y" | $tmp == "z") { return 9; }
	}
}

if (!function_exists('sync_directory')) {
	function sync_directory() {

		global $domain_uuid, $db;
		$settings_array = v_settings();
		foreach($settings_array as $name => $value) {
			$$name = $value;
		}

		$tmp = "include(\"config.js\");\n";
		$tmp .= "//var sounds_dir\n";
		$tmp .= "var admin_pin = \"\";\n";
		$tmp .= "var search_type = \"\";\n";
		$tmp .= "//var tmp_dir\n";
		$tmp .= "var digitmaxlength = 0;\n";
		$tmp .= "var timeoutpin = 5000;\n";
		$tmp .= "var timeouttransfer = 5000;\n";
		$tmp .= "\n";
		$tmp .= "var dtmf = new Object( );\n";
		$tmp .= "dtmf.digits = \"\";\n";
		$tmp .= "\n";
		$tmp .= "function mycb( session, type, obj, arg ) {\n";
		$tmp .= "	try {\n";
		$tmp .= "		if ( type == \"dtmf\" ) {\n";
		$tmp .= "			console_log( \"info\", \"digit: \"+obj.digit+\"\\n\" );\n";
		$tmp .= "			if ( obj.digit == \"#\" ) {\n";
		$tmp .= "				//console_log( \"info\", \"detected pound sign.\\n\" );\n";
		$tmp .= "				exit = true;\n";
		$tmp .= "				return( false );\n";
		$tmp .= "			}\n";
		$tmp .= "			if ( obj.digit == \"*\" ) {\n";
		$tmp .= "				//console_log( \"info\", \"detected pound sign.\\n\" );\n";
		$tmp .= "				exit = true;\n";
		$tmp .= "				return( false );\n";
		$tmp .= "			}\n";
		$tmp .= "			dtmf.digits += obj.digit;\n";
		$tmp .= "			if ( dtmf.digits.length >= digitmaxlength ) {\n";
		$tmp .= "				exit = true;\n";
		$tmp .= "				return( false );\n";
		$tmp .= "			}\n";
		$tmp .= "		}\n";
		$tmp .= "	} catch (e) {\n";
		$tmp .= "		console_log( \"err\", e+\"\\n\" );\n";
		$tmp .= "	}\n";
		$tmp .= "	return( true );\n";
		$tmp .= "} //end function mycb\n";
		$tmp .= "\n";
		$tmp .= "function directory_search(search_type) {\n";
		$tmp .= "\n";
		$tmp .= "	digitmaxlength = 3;\n";
		$tmp .= "	session.streamFile( sounds_dir+\"/en/us/callie/directory/48000/dir-enter_person.wav\");\n";
		$tmp .= "	if (search_type == \"last_name\") {\n";
		$tmp .= "		session.streamFile( sounds_dir+\"/en/us/callie/directory/48000/dir-last_name.wav\", mycb, \"dtmf\");\n";
		$tmp .= "		session.streamFile( sounds_dir+\"/en/us/callie/directory/48000/dir-to_search_by.wav\", mycb, \"dtmf\");\n";
		$tmp .= "		session.streamFile( sounds_dir+\"/en/us/callie/directory/48000/dir-first_name.wav\", mycb, \"dtmf\");\n";
		$tmp .= "	}\n";
		$tmp .= "	if (search_type == \"first_name\") {\n";
		$tmp .= "		session.streamFile( sounds_dir+\"/en/us/callie/directory/48000/dir-first_name.wav\", mycb, \"dtmf\");\n";
		$tmp .= "		session.streamFile( sounds_dir+\"/en/us/callie/directory/48000/dir-to_search_by.wav\", mycb, \"dtmf\");\n";
		$tmp .= "		session.streamFile( sounds_dir+\"/en/us/callie/directory/48000/dir-last_name.wav\", mycb, \"dtmf\");\n";
		$tmp .= "	}\n";
		$tmp .= "	session.streamFile( sounds_dir+\"/en/us/callie/directory/48000/dir-press.wav\", mycb, \"dtmf\");\n";
		$tmp .= "	session.execute(\"say\", \"en name_spelled iterated 1\");\n";
		$tmp .= "	session.collectInput( mycb, dtmf, timeoutpin );\n";
		$tmp .= "	var dtmf_search = dtmf.digits;\n";
		$tmp .= "	//console_log( \"info\", \"--\" + dtmf.digits + \"--\\n\" );\n";
		$tmp .= "	if (dtmf_search == \"1\") {\n";
		$tmp .= "		//console_log( \"info\", \"press 1 detected: \" + dtmf.digits + \"\\n\" );\n";
		$tmp .= "		//console_log( \"info\", \"press 1 detected: \" + search_type + \"\\n\" );\n";
		$tmp .= "		if (search_type == \"last_name\") {\n";
		$tmp .= "			//console_log( \"info\", \"press 1 detected last_name: \" + search_type + \"\\n\" );\n";
		$tmp .= "			search_type = \"first_name\";\n";
		$tmp .= "		}\n";
		$tmp .= "		else {\n";
		$tmp .= "			//console_log( \"info\", \"press 1 detected first_name: \" + search_type + \"\\n\" );\n";
		$tmp .= "			search_type = \"last_name\";\n";
		$tmp .= "		}\n";
		$tmp .= "		dtmf_search = \"\";\n";
		$tmp .= "		dtmf.digits = \"\";\n";
		$tmp .= "		directory_search(search_type);\n";
		$tmp .= "		return;\n";
		$tmp .= "	}\n";
		$tmp .= "	console_log( \"info\", \"first 3 letters of first or last name: \" + dtmf.digits + \"\\n\" );\n";
		$tmp .= "\n";
		$tmp .= "	//session.execute(\"say\", \"en name_spelled pronounced mark\");\n";
		$tmp .= "	//<action application=\"say\" data=\"en name_spelled iterated \${destination_number}\"/>\n";
		$tmp .= "	//session.execute(\"say\", \"en number iterated 12345\");\n";
		$tmp .= "	//session.execute(\"say\", \"en number pronounced 1001\");\n";
		$tmp .= "	//session.execute(\"say\", \"en short_date_time pronounced [timestamp]\");\n";
		$tmp .= "	//session.execute(\"say\", \"en CURRENT_TIME pronounced CURRENT_TIME\");\n";
		$tmp .= "	//session.execute(\"say\", \"en CURRENT_DATE pronounced CURRENT_DATE\");\n";
		$tmp .= "	//session.execute(\"say\", \"en CURRENT_DATE_TIME pronounced CURRENT_DATE_TIME\");\n";
		$tmp .= "\n";
		$tmp .= "\n";
		$tmp .= "	//take each name and convert it to the equivalent number in php when this file is generated\n";
		$tmp .= "	//then test each number see if it matches the user dtmf search keys\n";
		$tmp .= "\n";
		$tmp .= "	var result_array = new Array();\n";
		$tmp .= "	var x = 0;\n";

		//get a list of extensions and the users assigned to them
			$sql = "";
			$sql .= " select * from v_extensions ";
			$sql .= "where domain_uuid = '$domain_uuid' ";
			$prep_statement = $db->prepare(check_sql($sql));
			$prep_statement->execute();
			$x = 0;
			$result = $prep_statement->fetchAll(PDO::FETCH_NAMED);
			foreach ($result as &$row) {
				//print_r($row);
				$extension = $row["extension"];
				$effective_caller_id_name = $row["effective_caller_id_name"];
				//$user_list = $row["user_list"];
				//$user_list = trim($user_list, "|");
				//echo $user_list."<br />\n";
				//$username_array = explode ("|", $user_list);
				//print_r($username_array);
				foreach ($username_array as &$username) {
					if (strlen($username) > 0) {
						$sql = "";
						$sql .= "select * from v_users ";
						$sql .= "where domain_uuid = '$domain_uuid' ";
						$sql .= "and username = '$username' ";
						$prep_statement = $db->prepare(check_sql($sql));
						$prep_statement->execute();
						$tmp_result = $prep_statement->fetchAll(PDO::FETCH_NAMED);
						foreach ($tmp_result as &$row_tmp) {
							$user_first_name = $row_tmp["user_first_name"];
							$user_last_name = $row_tmp["user_last_name"];
							if ($user_first_name == "na") { $user_first_name = ""; }
							if ($user_last_name == "na") { $user_last_name = ""; }
							if (strlen($user_first_name.$user_last_name) == 0) {
								$name_array = explode (" ", $effective_caller_id_name);
								$user_first_name = $name_array[0];
								if (count($name_array) > 1) {
									$user_last_name = $name_array[1];
								}
							}
							
							break; //limit to 1 row
						}
						$f1 = phone_letter_to_number(substr($user_first_name, 0,1)); 
						$f2 = phone_letter_to_number(substr($user_first_name, 1,1));
						$f3 = phone_letter_to_number(substr($user_first_name, 2,1));

						$l1 = phone_letter_to_number(substr($user_last_name, 0,1)); 
						$l2 = phone_letter_to_number(substr($user_last_name, 1,1));
						$l3 = phone_letter_to_number(substr($user_last_name, 2,1));

						//echo $sql." extension: $extension  firstname $user_first_name lastname $user_last_name $tmp<br />";

						$tmp .= "	if (search_type == \"first_name\" && dtmf_search == \"".$f1.$f2.$f3."\" || search_type == \"last_name\" && dtmf_search == \"".$l1.$l2.$l3."\") {\n";
						$tmp .= "		result_array[x]=new Array()\n";
						$tmp .= "		result_array[x]['first_name'] =\"".$user_first_name."\";\n";
						$tmp .= "		result_array[x]['last_name'] =\"".$user_last_name."\";\n";
						$tmp .= "		result_array[x]['extension'] = \"".$extension."\";\n";
						$tmp .= "		//console_log( \"info\", \"found: ".$user_first_name." ".$user_last_name."\\n\" );\n";
						$tmp .= "		x++;\n";
						$tmp .= "	}\n";
					}
				}
			}
			unset ($prep_statement);

		$tmp .= "\n";
		$tmp .= "\n";
		$tmp .= "	//say the number of results that matched\n";
		$tmp .= "	\$result_count = result_array.length;\n";
		$tmp .= "	session.execute(\"say\", \"en number iterated \"+\$result_count);\n";
		$tmp .= "	session.streamFile( sounds_dir+\"/en/us/callie/directory/48000/dir-result_match.wav\", mycb, \"dtmf\");\n";
		$tmp .= "\n";
		$tmp .= "	//clear values\n";
		$tmp .= "	dtmf_search = 0;\n";
		$tmp .= "	dtmf.digits = '';\n";
		$tmp .= "\n";
		$tmp .= "	if (\$result_count == 0) {\n";
		$tmp .= "		//session.execute(\"transfer\", \"*347 XML default\");\n";
		$tmp .= "		directory_search(search_type);\n";
		$tmp .= "		return;\n";
		$tmp .= "	}\n";
		$tmp .= "\n";
		$tmp .= "	session.execute(\"set\", \"tts_engine=flite\");\n";
		$tmp .= "	session.execute(\"set\", \"tts_voice=rms\");  //rms //kal //awb //slt\n";
		$tmp .= "	session.execute(\"set\", \"playback_terminators=#\");\n";
		$tmp .= "	//session.speak(\"flite\",\"kal\",\"Thanks for.. calling\");\n";
		$tmp .= "\n";
		$tmp .= "	i=1;\n";
		$tmp .= "	for ( i in result_array ) {\n";
		$tmp .= "\n";
		$tmp .= "		//say first name and last name is at extension 1001\n";
		$tmp .= "		//session.execute(\"speak\", result_array[i]['first_name']);\n";
		$tmp .= "		//session.execute(\"speak\", result_array[i]['last_name']);\n";
		$tmp .= "		session.execute(\"say\", \"en name_spelled pronounced \"+result_array[i]['first_name']);\n";
		$tmp .= "		session.execute(\"sleep\", \"500\");\n";
		$tmp .= "		session.execute(\"say\", \"en name_spelled pronounced \"+result_array[i]['last_name']);\n";
		$tmp .= "		session.streamFile( sounds_dir+\"/en/us/callie/directory/48000/dir-at_extension.wav\", mycb, \"dtmf\");\n";
		$tmp .= "		session.execute(\"say\", \"en number pronounced \"+result_array[i]['extension']);\n";
		$tmp .= "\n";
		$tmp .= "		//to select this entry press 1\n";
		$tmp .= "		session.streamFile( sounds_dir+\"/en/us/callie/directory/48000/dir-to_select_entry.wav\", mycb, \"dtmf\");\n";
		$tmp .= "		session.streamFile( sounds_dir+\"/en/us/callie/directory/48000/dir-press.wav\", mycb, \"dtmf\");\n";
		$tmp .= "		session.execute(\"say\", \"en number iterated 1\");\n";
		$tmp .= "\n";
		$tmp .= "		//console_log( \"info\", \"first name: \" + result_array[i]['first_name'] + \"\\n\" );\n";
		$tmp .= "		//console_log( \"info\", \"last name: \" + result_array[i]['last_name'] + \"\\n\" );\n";
		$tmp .= "		//console_log( \"info\", \"extension: \" + result_array[i]['extension'] + \"\\n\" );\n";
		$tmp .= "\n";
		$tmp .= "		//if 1 is pressed then transfer the call\n";
		$tmp .= "		dtmf.digits = session.getDigits(1, \"#\", 3000);\n";
		$tmp .= "		if (dtmf.digits == \"1\") {\n";
		$tmp .= "			console_log( \"info\", \"directory: call transfered to: \" + result_array[i]['extension'] + \"\\n\" );\n";
		$tmp .= "			session.execute(\"transfer\", result_array[i]['extension']+\" XML default\");\n";
		$tmp .= "		}\n";
		$tmp .= "\n";
		$tmp .= "	}\n";
		$tmp .= "}\n";
		$tmp .= "\n";
		$tmp .= "\n";
		$tmp .= "if ( session.ready() ) {\n";
		$tmp .= "	session.answer();\n";
		$tmp .= "	search_type = \"last_name\";\n";
		$tmp .= "	directory_search(search_type);\n";
		$tmp .= "	session.hangup(\"NORMAL_CLEARING\");\n";
		$tmp .= "}\n";
		$tmp .= "";

		//write the file
		$fout = fopen($_SESSION['switch']['scripts']['dir']."/directory.js","w");
		fwrite($fout, $tmp);
		fclose($fout);

	} //end sync_directory
} //end if function exists

sync_directory();
?>

File: resources/install/scripts/directory.lua
--	directory.lua
--	Part of FusionPBX
--	Copyright (C) 2012 Mark J Crane <markjcrane@fusionpbx.com>
--	All rights reserved.
--
--	Redistribution and use in source and binary forms, with or without
--	modification, are permitted provided that the following conditions are met:
--
--	1. Redistributions of source code must retain the above copyright notice,
--	this list of conditions and the following disclaimer.
--
--	2. Redistributions in binary form must reproduce the above copyright
--	notice, this list of conditions and the following disclaimer in the
--	documentation and/or other materials provided with the distribution.
--
--	THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
--	INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
--	AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
--	AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
--	OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
--	SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
--	INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
--	CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
--	ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
--	POSSIBILITY OF SUCH DAMAGE.

--set the defaults
	digit_max_length = 3;
	timeout_pin = 5000;
	timeout_transfer = 5000;
	max_tries = 3;
	digit_timeout = 5000;
	search_limit = 3;
	search_count = 0;

--include config.lua
	require "resources.functions.config";

--connect to the database
	require "resources.functions.database_handle";
	dbh = database_handle('system');

--include functions
	require "resources.functions.format_ringback"

--settings
	require "resources.functions.settings";
	settings = settings(domain_uuid);
	storage_type = "";
	storage_path = "";
	if (settings['voicemail'] ~= nil) then
		if (settings['voicemail']['storage_type'] ~= nil) then
			if (settings['voicemail']['storage_type']['text'] ~= nil) then
				storage_type = settings['voicemail']['storage_type']['text'];
			end
		end
		if (settings['voicemail']['storage_path'] ~= nil) then
			if (settings['voicemail']['storage_path']['text'] ~= nil) then
				storage_path = settings['voicemail']['storage_path']['text'];
				storage_path = storage_path:gsub("${domain_name}", domain_name);
				storage_path = storage_path:gsub("${voicemail_id}", voicemail_id);
				storage_path = storage_path:gsub("${voicemail_dir}", voicemail_dir);
			end
		end
	end

--debug
	debug["info"] = false;
	debug["sql"] = false;

--prepare the api object
	api = freeswitch.API();

--get the session variables
	if ( session:ready() ) then
		--answer the session
			session:answer();

		--give time for the call to be ready
			session:streamFile("silence_stream://1000");

		--get the domain info
			domain_name = session:getVariable("domain_name");
			domain_uuid = session:getVariable("domain_uuid");

		--set the sounds path for the language, dialect and voice
			default_language = session:getVariable("default_language");
			default_dialect = session:getVariable("default_dialect");
			default_voice = session:getVariable("default_voice");
			if (not default_language) then default_language = 'en'; end
			if (not default_dialect) then default_dialect = 'us'; end
			if (not default_voice) then default_voice = 'callie'; end

		--set ringback
			directory_ringback = format_ringback(session:getVariable("ringback"));
			session:setVariable("ringback", directory_ringback);
			session:setVariable("transfer_ringback", directory_ringback);

		--set the sounds path for the language, dialect and voice
			session:setVariable("instant_ringback", "true");
			session:setVariable("ignore_early_media", "true");

		--define the sounds directory
			sounds_dir = session:getVariable("sounds_dir");
			sounds_dir = sounds_dir.."/"..default_language.."/"..default_dialect.."/"..default_voice;
	end

--set the voicemail_dir
	voicemail_dir = settings['switch']['voicemail']['dir'].."/default/"..domain_name;
	if (debug["info"]) then
		freeswitch.consoleLog("notice", "[directory] voicemail_dir: " .. voicemail_dir .. "\n");
	end

--get the domain_uuid
	if (domain_uuid == nil) then
		if (domain_name ~= nil) then
			sql = "SELECT domain_uuid FROM v_domains ";
			sql = sql .. "WHERE domain_name = '" .. domain_name .."' ";
			if (debug["sql"]) then
				freeswitch.consoleLog("notice", "[conference] SQL: " .. sql .. "\n");
			end
			status = dbh:query(sql, function(rows)
				domain_uuid = string.lower(rows["domain_uuid"]);
				end);
		end
	end

--define the explode function
	require "resources.functions.explode"

--define a function to convert dialpad letters to numbers
	function dialpad_to_digit(letter)
		letter = string.lower(letter);
		if (letter == "a" or letter == "b" or letter == "c") then
			digit = "2";
		elseif (letter == "d" or letter == "e" or letter == "f") then
			digit = "3";
		elseif (letter == "g" or letter == "h" or letter == "i") then
			digit = "4";
		elseif (letter == "j" or letter == "k" or letter == "l") then
			digit = "5";
		elseif (letter == "m" or letter == "n" or letter == "o") then
			digit = "6";
		elseif (letter == "p" or letter == "q" or letter == "r" or letter == "s") then
			digit = "7";
		elseif (letter == "t" or letter == "u" or letter == "v") then
			digit = "8";
		elseif (letter == "w" or letter == "x" or letter == "y" or letter == "z") then
			digit = "9";
		end
		return digit;
	end
	--print(dialpad_to_digit("m"));

--define table_count
	function table_count(T)
		local count = 0
		for _ in pairs(T) do count = count + 1 end
		return count
	end

--define the trim function
	require "resources.functions.trim"

--check if a file exists
	require "resources.functions.file_exists"

--define select_entry function 
	function select_entry()
		dtmf_digits = "";
		digit_timeout = "500";
		max_digits = 1;
		max_tries = 1;
		dtmf_digits = session:playAndGetDigits(min_digits, max_digits, max_tries, digit_timeout, "#", sounds_dir.."/directory/dir-to_select_entry.wav", "", "\\d+");
		if (string.len(dtmf_digits) == 0) then
			dtmf_digits = session:playAndGetDigits(min_digits, max_digits, max_tries, digit_timeout, "#", sounds_dir.."/voicemail/vm-press.wav", "", "\\d+");
		end
		if (string.len(dtmf_digits) == 0) then
			digit_timeout = "3000";
			dtmf_digits = session:playAndGetDigits(min_digits, max_digits, max_tries, digit_timeout, "#", sounds_dir.."/digits/1.wav", "", "\\d+");
		end
		return dtmf_digits;
	end

--define prompt_for_name function 
	function prompt_for_name()
		dtmf_digits = "";
		min_digits=0; max_digits=3; max_tries=3; digit_timeout = "5000";
		dtmf_digits = session:playAndGetDigits(min_digits, max_digits, max_tries, digit_timeout, "#", sounds_dir.."/directory/dir-enter_person_first_or_last.wav", "", "\\d+");
		return dtmf_digits;
	end

--define the directory_search function
	function directory_search()

		--get the digits for the name
			dtmf_digits = prompt_for_name();

		--show the dtmf digits 
			freeswitch.consoleLog("notice", "[directory] first 3 letters of first or last name: " .. dtmf_digits .. "\n");

		--loop through the extensions to find matches
			search_dtmf_digits = dtmf_digits;
			found = false;
			for key,row in pairs(directory) do

				--if (row.first_name and row.last_name) then
				--	freeswitch.consoleLog("notice", "[directory] ext: " .. row.extension .. " context " .. row.context .. " name " .. row.first_name .. " "..row.first_name_digits.." ".. row.last_name .. " "..row.last_name_digits.." "..row.directory_exten_visible.."\n");
				--else
				--	freeswitch.consoleLog("notice", "[directory] ext: " .. row.extension .. " context " .. row.context .. "\n");
				--end

				if (search_dtmf_digits == row.last_name_digits) or (search_dtmf_digits == row.first_name_digits) then
					if (row.first_name) then
						--play the recorded name
							if (storage_type == "base64") then
								sql = [[SELECT * FROM v_voicemails 
									WHERE domain_uuid = ']] .. domain_uuid ..[['
									AND voicemail_id = ']].. row.extension.. [[' ]];
								if (debug["sql"]) then
									freeswitch.consoleLog("notice", "[directory] SQL: " .. sql .. "\n");
								end
								status = dbh:query(sql, function(field)
									--add functions
										require "resources.functions.base64";

									--set the voicemail message path
										file_location = voicemail_dir.."/"..row.extension.."/recorded_name.wav";

									--save the recording to the file system
										if (string.len(field["voicemail_name_base64"]) > 32) then
											local file = io.open(file_location, "w");
											file:write(base64.decode(field["voicemail_name_base64"]));
											file:close();
										end

									--play the recorded name
										if (file_exists(file_location)) then
											session:streamFile(file_location);
										else
											--announce the first and last names
											session:execute("say", "en name_spelled iterated "..row.first_name);
											--session:execute("sleep", "500");
											if (row.last_name ~= nil) then
												session:execute("say", "en name_spelled iterated "..row.last_name);
											end
										end
								end);
							elseif (storage_type == "http_cache") then
								file_location = storage_path.."/"..row.extension.."/recorded_name.wav";
								if (file_exists(file_location)) then
									session:streamFile(file_location);
								end
							else
								if (debug["info"]) then
									freeswitch.consoleLog("notice", "[directory] path: "..voicemail_dir.."/"..row.extension.."/recorded_name.wav\n");
								end
								if (file_exists(voicemail_dir.."/"..row.extension.."/recorded_name.wav")) then
									session:streamFile(voicemail_dir.."/"..row.extension.."/recorded_name.wav");
								else
									--announce the first and last names
										session:execute("say", "en name_spelled iterated "..row.first_name);
										if (row.last_name ~= nil) then
											--session:execute("sleep", "500");
											session:execute("say", "en name_spelled iterated "..row.last_name);
										end
								end
							end

						--announce the extension number
							if (row.directory_exten_visible == "false") then
								--invisible extension number
							else
								session:streamFile(sounds_dir.."/directory/dir-at_extension.wav");
								session:execute("say", "en number iterated "..row.extension);
							end

						--select this entry press 1
							dtmf_digits = select_entry();

						--if 1 is pressed then transfer the call
							if (dtmf_digits == "1") then
								session:execute("transfer", row.extension.." XML "..row.context);
							end
					end
					found = true;
				end
			end
			if (found ~= true) then
				session:streamFile(sounds_dir.."/directory/dir-no_matching_results.wav");
			end
			search_count = search_count + 1;
			if (search_count < search_limit) then
				directory_search();
			end
	end

--get the extensions from the database
	sql = "SELECT * FROM v_extensions WHERE domain_uuid = '" .. domain_uuid .. "' AND enabled = 'true' AND (directory_visible is null or directory_visible = 'true'); ";
	if (debug["sql"]) then
		freeswitch.consoleLog("notice", "[directory] SQL: " .. sql .. "\n");
	end
	x = 1;
	directory = {}
	dbh:query(sql, function(row)
		--show all key value pairs
			--for key, val in pairs(row) do
			--	freeswitch.consoleLog("notice", "[directory] Key: " .. key .. " Value: " .. val .. "\n");
			--end
		--add the entire row to the directory table array
			--directory[x] = row;
		--variables
			effective_caller_id_name = row.effective_caller_id_name;
			if (row.directory_full_name) then
				name = row.directory_full_name;
			else
				if (string.len(effective_caller_id_name) > 0) then
					name = effective_caller_id_name;
				end
			end
			if (name) then
				name_table = explode(" ",name);
				first_name = name_table[1];
				last_name = name_table[2];
				if (first_name) then
					if (string.len(first_name) > 0) then
						--freeswitch.consoleLog("notice", "[directory] first_name: --" .. first_name .. "--\n");
						first_name_digits = dialpad_to_digit(string.sub(first_name, 1, 1))..dialpad_to_digit(string.sub(first_name, 2, 2))..dialpad_to_digit(string.sub(first_name, 3, 3));
					end
				end
				if (last_name) then
					if (string.len(last_name) > 0) then
						--freeswitch.consoleLog("notice", "[directory] last_name: --" .. last_name .. "--\n");
						last_name_digits = dialpad_to_digit(string.sub(last_name, 1, 1))..dialpad_to_digit(string.sub(last_name, 2, 2))..dialpad_to_digit(string.sub(last_name, 3, 3));
					end
				end

			end
		--add the row to the array
			--freeswitch.consoleLog("notice", "[directory] extension="..row.extension..",context="..row.user_context..",first_name="..name_table[1]..",last_name="..name_table[2]..",first_name_digits="..first_name_digits..",last_name_digits="..last_name_digits..",directory_exten_visible="..row.directory_exten_visible.."\n");
			table.insert(directory, {extension=row.extension,context=row.user_context,first_name=name_table[1],last_name=name_table[2],first_name_digits=first_name_digits,last_name_digits=last_name_digits,directory_exten_visible=row.directory_exten_visible});

		--increment x
			x = x + 1;
	end);

--call the directory search function
	if (session:ready()) then
		directory_search();
	end

	session:streamFile(sounds_dir.."/voicemail/vm-goodbye.wav");

--notes
	--session:execute("say", "en name_spelled pronounced mark");
	--<action application="say" data="en name_spelled iterated ${destination_number}"/>
	--session:execute("say", "en number iterated 12345");
	--session:execute("say", "en number pronounced 1001");
	--session:execute("say", "en short_date_time pronounced [timestamp]");
	--session:execute("say", "en CURRENT_TIME pronounced CURRENT_TIME");
	--session:execute("say", "en CURRENT_DATE pronounced CURRENT_DATE");
	--session:execute("say", "en CURRENT_DATE_TIME pronounced CURRENT_DATE_TIME");
